syntax = "proto3";

service Service {
    rpc CreateUser(User) returns (User);
    rpc GetUser(User) returns (User);
    rpc UpdateUser(User) returns (User);
    rpc DeleteUser(UserID) returns (Empty);

    rpc CreateEvent(Event) returns (Event);
    rpc GetEvent(EventID) returns (Event);
    rpc UpdateEvent(Event) returns (Event);
    rpc DeleteEvent(EventID) returns (Empty);
    rpc GetEventByUser(UserID) returns (stream Event);

    rpc CreateQuery(Query) returns (Query);
    rpc GetQuery(QueryID) returns (Query);
    rpc UpdateQuery(Query) returns (Query);
    rpc DeleteQuery(QueryID) returns (Empty);

    rpc CreateTweet(Tweet) returns (Tweet);
    rpc GetTweet(TweetID) returns (Tweet);
    rpc UpdateTweet(Tweet) returns (Tweet);
    rpc DeleteTweet(TweetID) returns (Empty);
    rpc SearchTweet(Filters) returns (stream Tweet);
}
message Empty{}

message User {
    int32 ID = 1;
    string name = 2;
    string password = 3;
    string content = 4;
}

message UserID{
    int32 ID = 1;
}

message Event {
    int32 ID = 1;
    string location = 2;
    string time = 3;
    string content = 4;
    int32 userID = 5;
}

message EventID{
    int32 ID = 1;
}

message Query {
    int32 ID = 1;
    string content = 2;
    int32 eventID = 3;
}

message QueryID{
    int32 ID = 1;
}

message Tweet {
    int32 ID = 1;
    string location = 2;
    string time = 3;
    string content = 4;
    int32 personID = 5;
}

message TweetID{
    int32 ID = 1;
}

message Filters{
    string keyword = 1;
    string media_type = 2;
    string since = 3;
    string until = 4;
    string language = 5;
    bool repost = 6;
    float latitude = 7;
    float longtitude = 8;
    float radius = 9;
    string radius_unit = 10;
    int32 min_retweet = 11;
    int32 min_fac = 12;
}
